File: prolog/triangle/triangle.pl
triangle(Side, Side, Side, "equilateral") :-
  valid_triangle(Side, Side, Side), !.

triangle(Side1, Side1, Side2, "isoceles") :-
  valid_triangle(Side1, Side1, Side2), !.

triangle(Side1, Side2, Side1, "isoceles") :-
  valid_triangle(Side1, Side1, Side2), !.

triangle(Side2, Side1, Side1, "isoceles") :-
  valid_triangle(Side1, Side1, Side2), !.

triangle(Side1, Side2, Side3, "scalene") :-
  valid_triangle(Side1, Side2, Side3).

valid_triangle(A, B, C) :-
  AB is A + B,
  AC is A + C,
  BC is B + C,
  not(A > BC),
  not(B > AC),
  not(C > AB),
  A \== 0,
  B \== 0,
  C \== 0.

File: haskell/triangle/src/Triangle.hs
module Triangle (TriangleType(..), triangleType) where

data TriangleType = Equilateral
                  | Isosceles
                  | Scalene
                  | Illegal
                  deriving (Eq, Show)

validTriangle :: (Num a, Ord a) => a -> a -> a -> Bool
validTriangle 0 _ _ = False
validTriangle _ 0 _ = False
validTriangle _ _ 0 = False
validTriangle a b c = a < bc && b < ac && c < ab where
  ab = a + b
  ac = a + c
  bc = b + c

triangleType :: (Num a, Ord a) => a -> a -> a -> TriangleType
triangleType a b c
  | not (validTriangle a b c)  = Illegal
  | a == b && b == c           = Equilateral
  | a /= b && b /= c && a /= c = Scalene
  | otherwise                  = Isosceles
